### PR TITLE
Tutorial Notebook: Using the PEFT library with a DNA Language Model.

### PR DESCRIPTION
**Pull request by rahulbshrestha**
This PR adds a tutorial notebook that uses the PEFT library with a DNA Language Model to solve a downstream task.

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
Thanks, this looks good so far. Ping me once the notebook is ready for review.

For general context, check #1837.

**Comment by rahulbshrestha** [comment]
@BenjaminBossan this tutorial is ready for review!

File: examples/dna_language_models/dna_lm.ipynb [hunk 1]
```diff
@@ -0,0 +1,1369 @@
...
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from transformers import Trainer, TrainingArguments\n",
+    "\n",
+    "# Define training arguments\n",
+    "training_args = TrainingArguments(\n",
+    "    output_dir='./results',\n",
+    "    evaluation_strategy=\"epoch\",\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:671]: We get a warning that `evaluation_strategy` is deprecated and `eval_strategy` should be used instead. Let's switch.

File: examples/dna_language_models/dna_lm.ipynb [hunk 2]
```diff
@@ -0,0 +1,1369 @@
...
+    "# Step 3: Evaluate the Predictions (Optional)\n",
+    "from sklearn.metrics import accuracy_score, precision_recall_fscore_support\n",
+    "\n",
+    "accuracy = accuracy_score(test_labels, predicted_labels)\n",
+    "print(f\"Accuracy: {accuracy}\")\n",
+    "\n",
+    "precision, recall, f1, _ = precision_recall_fscore_support(test_labels, predicted_labels, average='weighted')\n",
+    "print(f\"Precision: {precision}\")\n",
+    "print(f\"Recall: {recall}\")\n",
+    "print(f\"F1-Score: {f1}\")"
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:766]: Precision/recall/F1 are all the same and identical to accuracy. I wonder if it's not enough to just report accuracy? As accuracy can be calculated easily, this would also allow us to remove the sklearn dependency, which is also not a PEFT dependency and would otherwise require users to install it.

File: examples/dna_language_models/dna_lm.ipynb [hunk 3]
```diff
@@ -0,0 +1,1369 @@
...
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accuracy: 0.795\n",
+      "Precision: 0.794957729468599\n",
+      "Recall: 0.795\n",
+      "F1-Score: 0.794799276945093\n"
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:752]: Could please add a sentence or two about what we should think about the results. Is this good, bad, are there maybe published values that this could be compared to? If the result is not so good, maybe add a sentence along the lines of: "The results are not very good but this is because we train with little data and haven't spent time tuning the model for better results...".

File: examples/dna_language_models/dna_lm.ipynb [hunk 4]
```diff
@@ -0,0 +1,1369 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "id": "71b8a749-461e-4533-b1d0-cebc924d3dc0",
+   "metadata": {},
+   "source": [
+    "The LoraConfig object is instantiated to configure the PEFT parameters:\n",
+    "\n",
+    "- task_type: Specifies the type of task, in this case, sequence classification (SEQ_CLS).\n",
+    "- r: Number of re-parameterization steps.\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:865]: Maybe better: The rank of the LoRA matrices.

File: examples/dna_language_models/dna_lm.ipynb [hunk 5]
```diff
@@ -0,0 +1,1369 @@
...
+   "cell_type": "code",
+   "execution_count": 116,
+   "id": "6c223937-86ea-42ef-991a-050f23b21ef9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from peft import LoraConfig, TaskType\n",
+    "\n",
+    "peft_config = LoraConfig(\n",
+    "    #task_type=\"SEQ_CLS\",\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:995]: Why is this commented out?

File: examples/dna_language_models/dna_lm.ipynb [hunk 6]
```diff
@@ -0,0 +1,1369 @@
...
+    "    r=8,\n",
+    "    lora_alpha=32,\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:997]: Typically, alpha is ~ 2x rank. Did you test that and found that 32 works better?

File: examples/dna_language_models/dna_lm.ipynb [hunk 7]
```diff
@@ -0,0 +1,1369 @@
...
+    "    target_modules=[\"query\", \"key\", \"value\"],\n",
+    "    lora_dropout=0.01,\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:999]: This is a small value for dropout. Did you find that it works well?

File: examples/dna_language_models/dna_lm.ipynb [hunk 8]
```diff
@@ -0,0 +1,1369 @@
...
+       "TrainOutput(global_step=130, training_loss=0.6522058927095853, metrics={'train_runtime': 48.5074, 'train_samples_per_second': 41.231, 'train_steps_per_second': 2.68, 'total_flos': 0.0, 'train_loss': 0.6522058927095853, 'epoch': 10.0})"
+      ]
+     },
+     "execution_count": 119,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "from transformers import Trainer, TrainingArguments\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:1244]: No need to import a second time.

File: examples/dna_language_models/dna_lm.ipynb [hunk 9]
```diff
@@ -0,0 +1,1369 @@
...
+    "\n",
+    "# Define training arguments\n",
+    "training_args = TrainingArguments(\n",
+    "    output_dir='./results',\n",
+    "    evaluation_strategy=\"epoch\",\n",
+    "    learning_rate=2e-5,\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:1250]: For LoRA training, we can often get away with a ~10x higher learning rate. Did you test that? Maybe we can get better test scores this way.

File: examples/dna_language_models/dna_lm.ipynb [hunk 10]
```diff
@@ -0,0 +1,1369 @@
...
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Accuracy: 0.65\n",
+      "Precision: 0.6567213280696427\n",
+      "Recall: 0.65\n",
+      "F1-Score: 0.6491592433189871\n"
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:1330]: Again, giving some context on these results and some ideas how to improve them would be nice.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks a lot for providing this detailed notebook on DNA-LM training with PEFT. I added a couple of comments, please check them out.

File: examples/dna_language_models/dna_lm.ipynb (hunk 5 above)
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:995]: Is this necessary when specifying the LoraConfig? It seems to work fine without specifying the task type.

File: examples/dna_language_models/dna_lm.ipynb (hunk 2 above)
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:766]: Yes, changed!

File: examples/dna_language_models/dna_lm.ipynb [hunk 11]
```diff
@@ -0,0 +1,1369 @@
...
+   "execution_count": 98,
+   "id": "efa9441d-f44c-4ca3-b24c-fa5c853896cd",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "test_sequences = []\n",
+    "train_sequences = []\n",
+    "val_sequences = []\n",
+    "\n",
+    "dataset_limit = 200\n",
```
[the remark this answers is not in the document]
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:374]: Yes, the entire datasets takes way longer.

File: examples/dna_language_models/dna_lm.ipynb [hunk 12]
```diff
@@ -0,0 +1,1369 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 97,
+   "id": "75f267a9-82d1-4343-982e-9b1ea542a330",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_kmers(seq, k=6, stride=1):\n",
```
[the remark this answers is not in the document]
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:359]: Added!

File: examples/dna_language_models/dna_lm.ipynb [hunk 13]
```diff
@@ -0,0 +1,1369 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 91,
+   "id": "e73fae58-03e9-4acc-b0fc-9bc810c7d366",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tokenizer = AutoTokenizer.from_pretrained(\"gagneurlab/SpeciesLM\", revision = \"downstream_species_lm\")\n",
```
[the remark this answers is not in the document]
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:83]: I added some information, is that enough?

File: examples/dna_language_models/dna_lm.ipynb (hunk 6 above)
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:517]: I haven't experimented with the review comments regarding changing hyperparameters for obtaining a higher score, are these necessary for a tutorial notebook?

File: examples/dna_language_models/dna_lm.ipynb (hunk 13 above)
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply examples/dna_language_models/dna_lm.ipynb:84]: Yes, that helps. Still I think many users could be reluctant to try the model, so if you know another model that could be used instead, it would be nice to make the switch and just add a note here to suggest the SpeciesLM model.

File: examples/dna_language_models/dna_lm.ipynb (hunk 11 above)
[the remark this answers is not in the document]
Comment (BenjaminBossan) [reply examples/dna_language_models/dna_lm.ipynb:227]: Okay, so let's add a comment that this is an optimization for speed, but for best results users should set this variable to `None` and use the whole dataset.

File: examples/dna_language_models/dna_lm.ipynb [hunk 14]
```diff
@@ -0,0 +1,642 @@
...
+    "import torch\n",
+    "from torch import nn\n",
+    "\n",
+    "class DNA_LM(nn.Module):\n",
+    "    def __init__(self, model, num_labels):\n",
+    "        super(DNA_LM, self).__init__()\n",
+    "        self.model = model\n",
+    "        self.classifier = nn.Linear(768, num_labels)\n",
+    "        self.in_features = 768\n",
+    "        self.out_features = 2\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:334]: Those are not really needed, are they?

File: examples/dna_language_models/dna_lm.ipynb (hunk 6 above)
Comment (BenjaminBossan) [reply examples/dna_language_models/dna_lm.ipynb:517]: No, not necessary at all. It's just that when an example deviates from the defaults, I wonder if the author made these choices because they found them to work better or if there is another reason. Same with the dropout rate below.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the updates. There are only a few minor issues left.

Also, could you please run the whole notebook before committing it, so that the outputs are directly visible to users without having to run the notebook first.

File: examples/dna_language_models/dna_lm.ipynb (hunk 13 above)
[the remark this answers is not in the document]
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb:84]: I haven't used any other models, so will stick to SpeciesLM for now, since it's the one I'm most familiar with. It also comes from a lab from my university (TU Munich) so I know it is safe.

File: examples/dna_language_models/dna_lm.ipynb (hunk 14 above)
Comment (rahulbshrestha) [reply examples/dna_language_models/dna_lm.ipynb]: Apparently, it's needed else I get thrown an error when using Trainer that self.in_features and self.out_features is missing. I've made it a bit cleaner so the numbers aren't hard-coded.

**Comment by rahulbshrestha** [comment]
@BenjaminBossan I just added the second review changes as you requested.

File: examples/dna_language_models/dna_lm.ipynb [hunk 15]
```diff
@@ -0,0 +1,2941 @@
...
+      "metadata": {
+        "id": "efa9441d-f44c-4ca3-b24c-fa5c853896cd"
+      },
+      "outputs": [],
+      "source": [
+        "test_sequences = []\n",
+        "train_sequences = []\n",
+        "val_sequences = []\n",
+        "\n",
+        "dataset_limit = 200\n",
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:482]: Could you please add a comment that this value should be set to `None` if users want to use the whole dataset?

File: examples/dna_language_models/dna_lm.ipynb [hunk 16]
```diff
@@ -0,0 +1,2941 @@
...
+        "        if labels is not None:\n",
+        "            loss_fct = nn.CrossEntropyLoss()\n",
+        "            loss = loss_fct(logits.view(-1, self.out_features), labels.view(-1))\n",
+        "\n",
+        "        return (loss, logits) if loss is not None else logits\n",
+        "\n",
+        "# Number of classes for your classification task\n",
+        "num_labels = 2\n",
+        "classification_model = DNA_LM(lm, num_labels)\n",
+        "classification_model.to('cuda')"
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:713]:
```suggestion
        "classification_model.to('cuda')";
```

File: examples/dna_language_models/dna_lm.ipynb [hunk 17]
```diff
@@ -0,0 +1,2941 @@
...
+          "execution_count": 40,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "# Number of classes for your classification task\n",
+        "num_labels = 2\n",
+        "classification_model = DNA_LM(lm, num_labels)\n",
+        "classification_model.to('cuda')"
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:1192]:
```suggestion
        "classification_model.to('cuda')";
```

File: examples/dna_language_models/dna_lm.ipynb [hunk 18]
```diff
@@ -0,0 +1,2941 @@
...
+              ")"
+            ]
+          },
+          "execution_count": 39,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "classification_model"
```
Comment (BenjaminBossan) [inline examples/dna_language_models/dna_lm.ipynb:1068]: Can be removed

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the updates, we're almost good. Just a few small things:

- Let's add a comment about the data size limit
- Let's not print the model so often, as it clutters the notebook and distracts from the important things
- It would be nice to have a short conclusion, e.g.: How much memory is saved with PEFT? If users want to improve the results, what could they test next?

**Comment by rahulbshrestha** [comment]
@BenjaminBossan ok, I made changes as you requested!

**Review by BenjaminBossan — APPROVED** [review]
Thanks for adding this notebook, the last changes LGTM.
